aconfig/aconfig.py

```diff
@@ -83,7 +86,35 @@ def __deepcopy__(self, memo):
         '''This enables deepcopy to successfully copy a Config object, despite
         the default value semantics
         '''
-        return AttributeAccessDict(copy.deepcopy(dict(self)))
+        return self.__class__(copy.deepcopy(dict(self)))
+
+
+class ImmutableAttributeAccessDict(AttributeAccessDict):
+    """This class subclasses AttributeAccessDict and removes the setters,
+    to allow the creation of immutable dicts.
+
+    Using inheritance this way allows the dicts to be recursively created via
+    AttributeAccessDict, while maintaining nested immutability.
+    """
+
+    def __init__(self, input_map, *_):
+        """See :func:`~aconfig.aconfig.AttributeAccessDict.__init__`"""
+        assert isinstance(input_map, dict), \
+            '`input_map` argument should be of type dict, but found type: <{0}>'.format(
+                type(input_map))
+        # 🌶️🌶️🌶️: we explicitly cast back down to `dict` for the immutable case
+        # If we were to build an immutable dict from the top-down, that would
+        # obviously fail.
+        input_map = dict(input_map)
+        # Invoke the AttributeAccessDict initializer
+        super().__init__(input_map)
+
+    def __setitem__(self, key, value):
+        raise TypeError("ImmutableAttributeAccessDict does not support item assignment")
+
+    def __setattr__(self, key, value):
+        raise TypeError("ImmutableAttributeAccessDict does not support item assignment")
```

Comment:
I think this should be an `AttributeError`:


```py
class Bif:
    __slots__ = ("bif", "buff")
    def __init__(self, bif, buff):
        self.bif = bif
        self.buff = buff

b = Bif(1, 2)
b.baz = 34
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Bif' object has no attribute 'baz'
```

Reply:
```suggestion
        raise AttributeError("ImmutableAttributeAccessDict does not support attribute assignment")
```

Reply:
oh good point